p2p/tpu.md

```diff
@@ -0,0 +1,234 @@
+  V1[Validator]
+  V2[Validator]
+end
+
+R1[Node]
+
+C1 -- RPC --> R1
+C2 -- TPU --> R1
+R1 -- TPU --> L
+C3 -- TPU --> L
+C4 -- TPU ---> V1
+V1 -- TPU\nTPUvote --> L
+V2 -- TPUvote --> L
+```
+
+In a Solana cluster, full nodes usually participate as TPU relayers, although not required.
+All validators included in the leader schedule should operate an open TPU service.
+
+Nodes attempt to forward transactions to the current leader's TPU endpoints.
+The node identities of the leader schedule are well known (runtime intrinsic),
+and their TPU endpoints are part of CRDS ContactInfo.
+
+## Classes
+
+The network uses separate classes of TPU traffic for QoS.
+
+- The **TPUvote** class is Tower BFT consensus messages (highest priority).
+- The default **TPU** class is regular user transactions, generated by validators.
+- The **TPUfwd** class is user transactions that the previous leader did not fully process.
+
+## TPU/QUIC protocol v1
+	
+TPU/QUIC relays transactions over the QUIC transport protocol.
+
+QUIC is defined by [RFC 9000] and runs on UDP networks.
+
+Data is encrypted using TLS 1.3, defined by [RFC 8446].
+
+  [RFC 8446]: https://www.rfc-editor.org/rfc/rfc8446.html
+  [RFC 9000]: https://www.rfc-editor.org/rfc/rfc9000.html
+
+### Connection Parameters
+
+**TLS ciphersuites**
+
+Refer to [RFC 8446 Section 9.1](https://datatracker.ietf.org/doc/html/rfc8446#section-9.1) for Mandatory-to-Implement Cipher Suites.
+
+Additional ciphersuites may be supported and advertised in cryptographic negotiation.
+
+**Send quota**
+
+In QUIC, clients are only permitted to send as much data as specified via quota messages.
+If any quota is violated, the server should close the connection.
+
+On connection creation, the server should advertise non-zero [`MAX_DATA`] and [`MAX_STREAMS`] quotas.
+Both quotas should be continually replenished.
+
+  [`MAX_DATA`]: https://www.rfc-editor.org/rfc/rfc9000.html#name-max_data-frames
+  [`MAX_STREAMS`]: https://www.rfc-editor.org/rfc/rfc9000.html#name-max_streams-frames
+
+### Streaming Protocol
+
+Transactions are transmitted via client-to-server unidirectional QUIC streams.
+Every stream contains exactly one serialized transaction.
+
+TPU/QUIC servers should accept stream data of up to the maximum transaction size (1232 bytes).
+
+#### Single-packet transactions
+
+One or more small transactions can be transmitted in a single QUIC 1-RTT packet.
+	
+**Example QUIC frame**
+
+```
+STREAM Frame {
+  Prefix (5) = 0b00001     # Prefix present on all STREAM frames
+  Offset Present (1) = 0   # Offset is absent for first (only) fragment
+  Length Present (1)       # Length should be absent if the tx is last in this packet
+  Fin (1) = 1              # End-of-stream is set for last (only) fragment
+
+  Stream ID (i)     # Stream ID specified by server (implied length)
+  [Length (i)]      # Length of the stream data that follows
+  Stream Data (..)  # Contains the transaction
+}
+```
+
+#### Transaction fragmentation
```

Comment:
afaik this isn't supported today?

Reply:
quinn (QUIC library) will fragment oversize transactions. The QUIC 1-RTT and stream headers also eat away from available TPU, so this is probably already in use for max size txs.

Reply:
Resolving this, feel free to reopen if you disagree